test/pylib/log_browsing.py

```diff
@@ -0,0 +1,89 @@
+class ScyllaLogFile():
+    def __init__(self, thread_pool: ThreadPoolExecutor, logfile_path: str):
+        self.thread_pool = thread_pool # used for asynchronous IO operations
+        self.file = logfile_path
+        if not os.path.isfile(self.file):
+            pytest.fail("Log file {} does not exist".format(self.file))
+
+    # Returns "a mark" to the current position of this node Scylla log.
+    # This is for use with the from_mark parameter of watch_log_for method,
+    # allowing to watch the log from the position when this method was called.
+    def mark(self) -> int:
+        with open(self.file, 'r') as f:
+            f.seek(0, os.SEEK_END)
+            return f.tell()
+
+    # wait_for_log() checks if the log contains the given message.
+    # Because it may take time for the log message to be flushed, and sometimes
+    # we may want to look for messages about various delayed events, this
+    # function doesn't give up when it reaches the end of file, and rather
+    # retries until a given timeout. The timeout may be long, because successful
+    # tests will not wait for it. Note, however, that long timeouts will make
+    # xfailing tests slow.
+    # The timeout is in seconds.
+    # If from_mark is given, the log is searched from that position, otherwise
+    # from the beginning.
+    async def wait_for(self, pattern: str | re.Pattern, from_mark: Optional[int] = None, timeout: int = 600) -> None:
+        prog = re.compile(pattern)
+        with open(self.file, 'r') as f:
+            if from_mark is not None:
+                f.seek(from_mark)
+
+            loop = asyncio.get_running_loop()
+
+            async with asyncio.timeout(timeout):
+                logger.debug("Waiting for log message: %s", pattern)
+                while True:
+                    line = await loop.run_in_executor(self.thread_pool, f.readline)
+                    if line and prog.search(line):
+                        logger.debug("Found log message: %s", line)
+                        return
+
+
+    # Returns a list of lines matching the regular expression in the Scylla log.
+    # The list contains tuples of (line, match), where line is the full line
+    # from the log file, and match is the re.Match object for the matching
+    # expression.
+    # If filter_expr is given, only lines which do not match it are returned.
+    # If from_mark is given, the log is searched from that position, otherwise
+    # from the beginning.
+    def grep(self, expr: str | re.Pattern, filter_expr: Optional[str | re.Pattern] = None,
+             from_mark: Optional[int] = None) -> list[(str, re.Match[str])]:
+        matchings = []
+        pattern = re.compile(expr)
+        if filter_expr:
+            filter_pattern = re.compile(filter_expr)
+        else:
+            filter_pattern = None
+        with open(self.file) as f:
+            if from_mark:
+                f.seek(from_mark)
+            for line in f:
```

Comment:
This is accessing the file synchronously.

Reply:
Is it a problem?